dependency_graph.py

```diff
@@ -22,32 +23,53 @@ def __eq__(self, other) -> bool:
             else:
                 return False
 
+
     def __init__(self):
         self.terminal_node = list()
 
+
+    @staticmethod
+    def from_circuit_by_commutation(circuit: Circuit) -> 'DependencyGraph':
+        """
+        Build a dependency tree from a Pauli circuit based on commutation.
+
+        """
+        # This because new dependency is added between non-commuting operations  
+        def func(arg1, arg2):
+            return not Circuit.are_commuting(arg1, arg2)
+        
+        return DependencyGraph.from_list(circuit.ops, func)
+
+
     @staticmethod
-    def from_circuit(circuit : Circuit) -> 'DependencyGraph':
+    def from_list(input_list: list(), comparing_function) -> 'DependencyGraph':
```

Comment:
Might be worth trying some other time:
```suggestion
    def from_list(input_list: list[T], comparing_function : Callable[T,T]) -> 'DependencyGraph[T]':
```